File: Drag&Drop/DragDropController/UICollectionView+CellSwapSupport.m
//
//  UICollectionView+CellSwapSupport.m
//  Drag&Drop
//
//  Created by Joey Patino on 11/17/15.
//  Copyright © 2015 Joseph Patino. All rights reserved.
//

#import <objc/runtime.h>

#import "UICollectionView+CellSwapSupport.h"
#import "NSIndexPath+Additions.h"

@interface UICollectionView ()
@end

@implementation UICollectionView (CellSwapSupport)

- (void)startCellSwapFromIndexPath:(NSIndexPath *)fromIndexPath
                  inCollectionView:(UICollectionView *)collectionView
                        toLocation:(CGPoint)location {
    DLog(@"%s", __PRETTY_FUNCTION__);

    self.cellSwapDestination = [self indexPathAtLocation:location];
    self.cellSwapOrigin = self.cellSwapDestination;

    if ([collectionView canMoveIndexPath:fromIndexPath to:self.cellSwapDestination withDestination:self]) {
        [collectionView moveIndexPath:fromIndexPath to:self.cellSwapDestination withDestination:self];

        [self insertVacancyAtIndexPath:self.cellSwapDestination animated:YES];
    }
    else {
        [self resetAfterSwap];
    }
}

- (void)continueCellSwap:(CGPoint)location {
    if (!self.cellSwapDestination && !self.cellSwapOrigin) return;
    
    self.cellSwapDestination = [self indexPathAtLocation:location];

    if (![self.cellSwapOrigin isIndexPath:self.cellSwapDestination]) {

        BOOL canMove = YES;
        if ([self.dataSource respondsToSelector:@selector(collectionView:canMoveItemAtIndexPath:)])
            canMove = [self.dataSource collectionView:self canMoveItemAtIndexPath:self.cellSwapOrigin];
        
        if (canMove && [self.dataSource respondsToSelector:@selector(collectionView:moveItemAtIndexPath:toIndexPath:)])
            [self.dataSource collectionView:self moveItemAtIndexPath:self.cellSwapOrigin toIndexPath:self.cellSwapDestination];

        [self insertVacancyAtIndexPath:self.cellSwapDestination animated:YES];
    }

    self.cellSwapOrigin = self.cellSwapDestination;
}


- (void)reverseCellSwapFromIndexPath:(NSIndexPath *)fromIndexPath inCollectionView:(UICollectionView *)collectionView {
    DLog(@"%s", __PRETTY_FUNCTION__);
    if (!self.cellSwapDestination && !self.cellSwapOrigin) return;

    [self layoutCollectionViewAnimated:YES completion:nil];

    if ([self canMoveIndexPath:self.cellSwapOrigin to:fromIndexPath withDestination:collectionView]) {
        [self moveIndexPath:self.cellSwapOrigin to:fromIndexPath withDestination:collectionView];
    }
}

#pragma mark -

- (void)insertCellAtIndexPath:(NSIndexPath *)indexPath {
    DLog(@"%s", __PRETTY_FUNCTION__);

    [UIView setAnimationsEnabled:NO];
    [self insertItemsAtIndexPaths:@[indexPath]];
    [self reloadItemsAtIndexPaths:[self indexPathsForVisibleItems]];
    [UIView setAnimationsEnabled:YES];
    
    [self resetAfterSwap];
}

- (void)deleteCellAtIndexPath:(NSIndexPath *)indexPath {
    [UIView setAnimationsEnabled:NO];
    [self deleteItemsAtIndexPaths:@[indexPath]];
    [self reloadItemsAtIndexPaths:[self indexPathsForVisibleItems]];
    [UIView setAnimationsEnabled:YES];
}

#pragma mark -

- (void)resetAfterSwap {
    DLog(@"%s", __PRETTY_FUNCTION__);

    self.cellSwapOrigin = nil;
    self.cellSwapDestination = nil;
}

#pragma mark - Cell Rearrangement

- (BOOL)shouldAcceptCellSwapFrom:(UICollectionView *)source {
    DLog(@"%s", __PRETTY_FUNCTION__);
    return (self.cellSwapOrigin && self.cellSwapDestination);
}

// Future support for selective cell swaps...
- (BOOL)canMoveIndexPath:(NSIndexPath *)fromIndexPath to:(NSIndexPath *)toIndexPath withDestination:(UICollectionView *)destination {
    
    NSObject <UICollectionViewDataSource_CellSwapSupport> *dataSource = (NSObject <UICollectionViewDataSource_CellSwapSupport> *)destination.dataSource;
    
    if ([dataSource respondsToSelector:@selector(collectionView:canMoveItemAtIndexPath:toCollectionView:toIndexPath:)])
        return [dataSource collectionView:self canMoveItemAtIndexPath:fromIndexPath toCollectionView:destination toIndexPath:toIndexPath];

    return YES;
}

- (void)moveIndexPath:(NSIndexPath *)fromIndexPath to:(NSIndexPath *)toIndexPath withDestination:(UICollectionView *)destination {

    NSObject <UICollectionViewDataSource_CellSwapSupport> *dataSource = (NSObject <UICollectionViewDataSource_CellSwapSupport> *)self.dataSource;
    
    if ([dataSource respondsToSelector:@selector(collectionView:moveItemAtIndexPath:toCollectionView:toIndexPath:)])
        [dataSource collectionView:self moveItemAtIndexPath:fromIndexPath toCollectionView:destination toIndexPath:toIndexPath];
}

- (void)insertVacancyAtIndexPath:(NSIndexPath *)toIndexPath animated:(BOOL)animated {
    DLog(@"%s", __PRETTY_FUNCTION__);

    [UIView animateWithDuration:animated ? .3 : 0.0 animations:^{
        
        [[self indexPathsForVisibleItems] enumerateObjectsUsingBlock:^(NSIndexPath *indexPath, NSUInteger idx, BOOL *stop){

            CGRect frame = [self.collectionViewLayout layoutAttributesForItemAtIndexPath:indexPath].frame;

            if ([indexPath isAfterIndexPath:toIndexPath] || [indexPath isIndexPath:toIndexPath])
                frame = [self.collectionViewLayout layoutAttributesForItemAtIndexPath:[indexPath indexPathByIncrementingRow]].frame;

            [self cellForItemAtIndexPath:indexPath].frame = frame;
        }];
        
    } completion:NULL];
}

- (void)layoutCollectionViewAnimated:(BOOL)animated completion:(dispatch_block_t)completion {
    DLog(@"%s", __PRETTY_FUNCTION__);

    [UIView animateWithDuration:animated ? .3 : 0.0 animations:^{
        
        [[self indexPathsForVisibleItems] enumerateObjectsUsingBlock:^(NSIndexPath *indexPath, NSUInteger idx, BOOL *stop){
            CGRect frame = [self.collectionViewLayout layoutAttributesForItemAtIndexPath:indexPath].frame;
            [self cellForItemAtIndexPath:indexPath].frame = frame;
        }];
        
    } completion:^(BOOL finished){
        if (completion)
            completion();
    }];
}

#pragma mark - Helpers

- (NSIndexPath *)indexPathAtLocation:(CGPoint)location {
    NSIndexPath *indexPath = [self indexPathForItemAtPoint:location];
    if (!indexPath) indexPath = [self findClosestIndexPathToPoint:location];
    return indexPath;
}

- (NSIndexPath *)findClosestIndexPathToPoint:(CGPoint)touchLocation {
    
    NSArray *collectionViewIndexPaths = [[self indexPathsForVisibleItems] sortedArrayUsingComparator:^NSComparisonResult(NSIndexPath *idx1, NSIndexPath *idx2){
        return [idx1 compare:idx2];
    }];
    
    CGFloat distanceToNearestIndexPath = CGFLOAT_MAX;
    BOOL locationIsBelowVisibleCells = YES;
    NSIndexPath *nearestIndexPath = nil;
    
    BOOL (^PointIsBelowFrame)(CGPoint, CGRect) = ^BOOL(CGPoint point, CGRect rect) {
        if (point.y <= CGRectGetMaxY(rect)) return YES;
        return NO;
    };
    
    BOOL (^PointIsToTheRightOfFrame)(CGPoint, CGRect) = ^BOOL(CGPoint point, CGRect rect) {
        if (point.x >= CGRectGetMaxX(rect)) return YES;
        return NO;
    };
    
    for (NSIndexPath *indexPath in collectionViewIndexPaths) {
        UICollectionViewLayoutAttributes *cellLayoutAttributes = [self.collectionViewLayout layoutAttributesForItemAtIndexPath:indexPath];
        CGFloat distanceToCell = [self distanceBetween:touchLocation p2:cellLayoutAttributes.center];
        
        if (PointIsBelowFrame(touchLocation, cellLayoutAttributes.frame)) {
            locationIsBelowVisibleCells = NO;
        }

        if ([collectionViewIndexPaths lastObject] == indexPath &&
            (PointIsToTheRightOfFrame(touchLocation, cellLayoutAttributes.frame))) {
            locationIsBelowVisibleCells = YES;
        }

        if (distanceToCell < distanceToNearestIndexPath) {
            distanceToNearestIndexPath = distanceToCell;
            nearestIndexPath = indexPath;
        }
    }
    
    if (locationIsBelowVisibleCells) {

        NSUInteger section = 0;
        NSUInteger row = 0;

        if (nearestIndexPath) {
            section = nearestIndexPath.section;
            row = [self.dataSource collectionView:self numberOfItemsInSection:section] - 1;
        }

        nearestIndexPath = [NSIndexPath indexPathForRow:row inSection:section];
    }
    
    return nearestIndexPath;
}

- (CGFloat)distanceBetween:(CGPoint)p1 p2:(CGPoint)p2 {
    
    CGFloat x1 = p1.x;
    CGFloat x2 = p2.x;
    
    CGFloat y1 = p1.y;
    CGFloat y2 = p2.y;
    
    double dx = (x2-x1);
    double dy = (y2-y1);
    double dist = sqrt(dx*dx + dy*dy);
    
    return dist;
}

#pragma mark -

- (void)setCellSwapOrigin:(NSIndexPath *)cellSwapOrigin {
    objc_setAssociatedObject(self, @selector(cellSwapOrigin), cellSwapOrigin, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

- (NSIndexPath *)cellSwapOrigin {
    return objc_getAssociatedObject(self, @selector(cellSwapOrigin));
}

- (void)setCellSwapDestination:(NSIndexPath *)cellSwapDestination {
    objc_setAssociatedObject(self, @selector(cellSwapDestination), cellSwapDestination, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

- (NSIndexPath *)cellSwapDestination {
    return objc_getAssociatedObject(self, @selector(cellSwapDestination));
}

@end

File: Sources/DragDrop/UICollectionView+CellSwapSupport.swift
//
//  UICollectionView+CellSwapSupport.swift
//  DragDrop
//
//  Created by Joey Patino on 11/17/15.
//  Copyright © 2015 Joseph Patino. All rights reserved.
//

import UIKit

/// Ported from UICollectionViewDataSource_CellSwapSupport. `respondsToSelector:`
/// against it becomes a conditional cast at the call sites.
@MainActor
public protocol UICollectionViewDataSourceCellSwapSupport: UICollectionViewDataSource {

    /// Was @optional. Defaults to true.
    func collectionView(_ sourceCollectionView: UICollectionView,
                        canMoveItemAt sourceIndexPath: IndexPath,
                        to destinationCollectionView: UICollectionView,
                        to destinationIndexPath: IndexPath) -> Bool

    /// Was @required.
    func collectionView(_ collectionView: UICollectionView,
                        moveItemAt sourceIndexPath: IndexPath,
                        to destinationCollectionView: UICollectionView,
                        to destinationIndexPath: IndexPath)
}

public extension UICollectionViewDataSourceCellSwapSupport {
    func collectionView(_ sourceCollectionView: UICollectionView,
                        canMoveItemAt sourceIndexPath: IndexPath,
                        to destinationCollectionView: UICollectionView,
                        to destinationIndexPath: IndexPath) -> Bool { true }
}

public extension UICollectionView {

    var cellSwapOrigin: IndexPath? {
        get { dragDropState.cellSwapOrigin }
        set { dragDropState.cellSwapOrigin = newValue }
    }

    var cellSwapDestination: IndexPath? {
        get { dragDropState.cellSwapDestination }
        set { dragDropState.cellSwapDestination = newValue }
    }

    // MARK: -

    func startCellSwap(from fromIndexPath: IndexPath?,
                       in collectionView: UICollectionView,
                       to location: CGPoint) {
        DLog()

        cellSwapDestination = indexPath(at: location)
        cellSwapOrigin = cellSwapDestination

        if let fromIndexPath, let destination = cellSwapDestination,
           collectionView.canMoveIndexPath(fromIndexPath, to: destination, withDestination: self) {
            collectionView.moveIndexPath(fromIndexPath, to: destination, withDestination: self)

            insertVacancy(at: destination, animated: true)
        } else {
            resetAfterSwap()
        }
    }

    func continueCellSwap(_ location: CGPoint) {
        if cellSwapDestination == nil && cellSwapOrigin == nil { return }

        cellSwapDestination = indexPath(at: location)

        if cellSwapOrigin?.isSame(as: cellSwapDestination ?? IndexPath(row: -1, section: -1)) != true {

            var canMove = true
            if let origin = cellSwapOrigin {
                canMove = dataSource?.collectionView?(self, canMoveItemAt: origin) ?? true
            }

            if canMove, let origin = cellSwapOrigin, let destination = cellSwapDestination {
                dataSource?.collectionView?(self, moveItemAt: origin, to: destination)
            }

            if let destination = cellSwapDestination {
                insertVacancy(at: destination, animated: true)
            }
        }

        cellSwapOrigin = cellSwapDestination
    }

    func reverseCellSwap(from fromIndexPath: IndexPath?, in collectionView: UICollectionView) {
        DLog()
        if cellSwapDestination == nil && cellSwapOrigin == nil { return }

        layoutCollectionView(animated: true, completion: nil)

        if let origin = cellSwapOrigin, let fromIndexPath,
           canMoveIndexPath(origin, to: fromIndexPath, withDestination: collectionView) {
            moveIndexPath(origin, to: fromIndexPath, withDestination: collectionView)
        }
    }

    // MARK: -

    func insertCell(at indexPath: IndexPath?) {
        DLog()
        guard let indexPath else { return }

        UIView.setAnimationsEnabled(false)
        insertItems(at: [indexPath])
        reloadItems(at: indexPathsForVisibleItems)
        UIView.setAnimationsEnabled(true)

        resetAfterSwap()
    }

    func deleteCell(at indexPath: IndexPath?) {
        guard let indexPath else { return }

        UIView.setAnimationsEnabled(false)
        deleteItems(at: [indexPath])
        reloadItems(at: indexPathsForVisibleItems)
        UIView.setAnimationsEnabled(true)
    }

    // MARK: -

    internal func resetAfterSwap() {
        DLog()

        cellSwapOrigin = nil
        cellSwapDestination = nil
    }

    // MARK: - Cell Rearrangement

    func shouldAcceptCellSwap(from source: UICollectionView) -> Bool {
        DLog()
        return cellSwapOrigin != nil && cellSwapDestination != nil
    }

    /// Future support for selective cell swaps...
    internal func canMoveIndexPath(_ fromIndexPath: IndexPath,
                                   to toIndexPath: IndexPath,
                                   withDestination destination: UICollectionView) -> Bool {

        guard let dataSource = destination.dataSource as? any UICollectionViewDataSourceCellSwapSupport else {
            return true
        }

        return dataSource.collectionView(self, canMoveItemAt: fromIndexPath,
                                         to: destination, to: toIndexPath)
    }

    internal func moveIndexPath(_ fromIndexPath: IndexPath,
                                to toIndexPath: IndexPath,
                                withDestination destination: UICollectionView) {

        guard let dataSource = self.dataSource as? any UICollectionViewDataSourceCellSwapSupport else {
            return
        }

        dataSource.collectionView(self, moveItemAt: fromIndexPath,
                                  to: destination, to: toIndexPath)
    }

    internal func insertVacancy(at toIndexPath: IndexPath, animated: Bool) {
        DLog()

        UIView.animate(withDuration: animated ? 0.3 : 0.0) { [self] in

            for indexPath in indexPathsForVisibleItems {

                var frame = frameForItem(at: indexPath)

                if indexPath.isAfter(toIndexPath) || indexPath.isSame(as: toIndexPath) {
                    frame = frameForItem(at: indexPath.incrementingRow)
                }

                cellForItem(at: indexPath)?.frame = frame
            }
        }
    }

    func layoutCollectionView(animated: Bool, completion: (() -> Void)?) {
        DLog()

        UIView.animate(withDuration: animated ? 0.3 : 0.0) { [self] in

            for indexPath in indexPathsForVisibleItems {
                cellForItem(at: indexPath)?.frame = frameForItem(at: indexPath)
            }

        } completion: { _ in
            completion?()
        }
    }

    // MARK: - Helpers

    internal func indexPath(at location: CGPoint) -> IndexPath? {
        indexPathForItem(at: location) ?? findClosestIndexPath(to: location)
    }

    internal func findClosestIndexPath(to touchLocation: CGPoint) -> IndexPath? {

        let collectionViewIndexPaths = indexPathsForVisibleItems.sorted { $0.compare($1) == .orderedAscending }

        var distanceToNearestIndexPath = CGFloat.greatestFiniteMagnitude
        var locationIsBelowVisibleCells = true
        var nearestIndexPath: IndexPath?

        let pointIsBelowFrame: (CGPoint, CGRect) -> Bool = { point, rect in
            point.y <= rect.maxY
        }

        let pointIsToTheRightOfFrame: (CGPoint, CGRect) -> Bool = { point, rect in
            point.x >= rect.maxX
        }

        for indexPath in collectionViewIndexPaths {
            guard let cellLayoutAttributes = collectionViewLayout.layoutAttributesForItem(at: indexPath) else {
                continue
            }
            let distanceToCell = distance(touchLocation, cellLayoutAttributes.center)

            if pointIsBelowFrame(touchLocation, cellLayoutAttributes.frame) {
                locationIsBelowVisibleCells = false
            }

            if collectionViewIndexPaths.last == indexPath
                && pointIsToTheRightOfFrame(touchLocation, cellLayoutAttributes.frame) {
                locationIsBelowVisibleCells = true
            }

            if distanceToCell < distanceToNearestIndexPath {
                distanceToNearestIndexPath = distanceToCell
                nearestIndexPath = indexPath
            }
        }

        if locationIsBelowVisibleCells {

            var section = 0
            var row = 0

            if let nearestIndexPath {
                section = nearestIndexPath.section
                row = (dataSource?.collectionView(self, numberOfItemsInSection: section) ?? 1) - 1
            }

            nearestIndexPath = IndexPath(row: row, section: section)
        }

        return nearestIndexPath
    }

    internal func distance(_ p1: CGPoint, _ p2: CGPoint) -> CGFloat {
        let dx = p2.x - p1.x
        let dy = p2.y - p1.y
        return sqrt(dx * dx + dy * dy)
    }
}
